From 1d53830a799d8ba639822490c22e8735e3a8a22f Mon Sep 17 00:00:00 2001
From: David Boddie <david.boddie@qt.io>
Date: Tue, 16 Dec 2025 12:02:14 +0100
Subject: QDoc: Issue warnings instead of fatal errors when @-file is invalid

If an include file is not specified, or does not correspond to a valid
file, QDoc issues a fatal warning, causing a core dump. This isn't very
useful for debugging and is inconsistent with other error reporting.

Instead, we now issue a warning and skip the invalid command line
argument.

Fixes: QTBUG-142742
Pick-to: 6.8
Change-Id: I27d1725d1ea20e51e1f5fefac1bf15d2be277d8a
Reviewed-by: Paul Wicking <paul.wicking@qt.io>
(cherry picked from commit 55ee9bacd08c632f9dd5492f5981a9fc5e571536)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
(cherry picked from commit 6e6a23b3fe4bd06af813d9c8619147af7b6c6652)
---
 src/qdoc/qdoc/src/qdoc/qdoccommandlineparser.cpp | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/qdoc/qdoc/src/qdoc/qdoccommandlineparser.cpp b/src/qdoc/qdoc/src/qdoc/qdoccommandlineparser.cpp
index b8705cac4..b60cb398c 100644
--- a/src/qdoc/qdoc/src/qdoc/qdoccommandlineparser.cpp
+++ b/src/qdoc/qdoc/src/qdoc/qdoccommandlineparser.cpp
@@ -150,12 +150,16 @@ static QStringList argumentsFromCommandLineAndFile(const QStringList &arguments)
         if (argument.startsWith(QLatin1Char('@'))) {
             QString optionsFile = argument;
             optionsFile.remove(0, 1);
-            if (optionsFile.isEmpty())
-                qFatal("The @ option requires an input file");
+            if (optionsFile.isEmpty()) {
+                qWarning("The @ option requires an input file");
+                continue;
+            }
             QFile f(optionsFile);
-            if (!f.open(QIODevice::ReadOnly | QIODevice::Text))
-                qFatal("Cannot open options file specified with @: %ls",
+            if (!f.open(QIODevice::ReadOnly | QIODevice::Text)) {
+                qWarning("Cannot open options file specified with @: %ls",
                        qUtf16Printable(optionsFile));
+                continue;
+            }
             while (!f.atEnd()) {
                 QString line = QString::fromLocal8Bit(f.readLine().trimmed());
                 if (!line.isEmpty())
-- 
2.49.0

